primitives/src/calendar.rs

```diff
@@ -0,0 +1,287 @@
+                let multiplier = u32::try_from(multiplier).ok()?;
+                let year_diff = u32::try_from(date_now.year() - year_start).ok()?;
+                // Convert months to base 12.
+                let month_start_base12 = month_start - 1;
+                let month_now_base12 = date_now.month() - 1;
+                let elapsed_months = year_diff * 12 + month_now_base12 - month_start_base12;
+                let elapsed_periods = elapsed_months / multiplier;
+                let next_period_months = multiplier * (elapsed_periods + 1);
+                // The month of the next period counting from the beginning of `year_start`.
+                let denormalized_next_period_month = month_start_base12 + next_period_months;
+                let next_period_year =
+                    year_start + i32::try_from(denormalized_next_period_month / 12).ok()?;
+                // Month in base 12.
+                let next_period_month = denormalized_next_period_month % 12;
+                let next_period_day = if matches!(next_period_month, 3 | 5 | 8 | 10) {
+                    // Handle 30-day months.
+                    u32::min(day_start, 30)
+                } else if next_period_month == 1 {
+                    // Handle February
+                    if is_leap_year(next_period_year) {
+                        u32::min(day_start, 29)
+                    } else {
+                        u32::min(day_start, 28)
+                    }
+                } else {
+                    day_start
+                };
+                let date_next = NaiveDate::from_ymd(
+                    i32::try_from(next_period_year).ok()?,
+                    // Convert months back to calendar numbering.
+                    next_period_month + 1,
+                    next_period_day,
+                );
+                Some(u64::try_from(date_next.and_time(date_time_start.time()).timestamp()).ok()?)
+            }
+            CalendarUnit::Year => {
+                // Convert the multiplier to match the type of year.
+                let multiplier = i32::try_from(multiplier).ok()?;
+                let elapsed_periods: i32 = (date_now.year() - year_start) / multiplier;
+                let next_period_year = year_start + multiplier * (elapsed_periods + 1);
+                let next_period_day = if month_start == 2 && !is_leap_year(next_period_year) {
+                    // Handle February in non-leap years.
+                    u32::min(day_start, 28)
+                } else {
+                    day_start
+                };
+                let date_next = NaiveDate::from_ymd(
+                    next_period_year,
+                    // Convert months back to calendar numbering.
+                    month_start,
+                    next_period_day,
+                );
+                Some(u64::try_from(date_next.and_time(date_time_start.time()).timestamp()).ok()?)
+            }
+            _ => unreachable!(),
```

Comment:
Runtime should never panic. even if this is unreachable, let's return something like `Some(very_large_num)` and document the behavior.

Reply:
I'll redo this at type level.